unixfs/src/file/reader.rs

```diff
@@ -0,0 +1,333 @@
+                return Err(FileError::UnexpectedRawOrFileProperties {
+                    hash_type: inner.data.hashType,
+                    fanout: inner.data.fanout,
+                }
+                .into());
+            }
+
+            let end = if inner.links.is_empty() {
+                // can unwrap because `data` is all of the data
+                let filesize = inner.data.filesize.unwrap_or(data.len() as u64);
+                Ending::Chunk(offset + filesize)
+            } else {
+                match inner.data.filesize {
+                    Some(filesize) => Ending::TreeCoverage(offset + filesize),
+                    None => return Err(FileError::IntermediateNodeWithoutFileSize.into()),
+                }
+            };
+
+            Ok(Self {
+                offset,
+                end,
+                had_links: !inner.links.is_empty(),
+                links: inner.links,
+                data,
+                blocksizes: inner.data.blocksizes,
+                metadata,
+            })
+        }
+    }
+
+    /// Returns a moved tuple of the content (bytes or links) and a traversal, which can be used to
+    /// continue the traversal from the next block.
+    pub fn content(
+        self,
+    ) -> (
+        FileContent<'a, impl Iterator<Item = (PBLink<'a>, Range<u64>)>>,
+        Traversal,
+    ) {
+        let traversal = Traversal {
+            last_had_links: !self.had_links,
+            last_ending: self.end,
+            last_offset: self.offset,
+
+            metadata: self.metadata,
+        };
+
+        if self.links.is_empty() {
+            (FileContent::Just(self.data), traversal)
+        } else {
+            let zipped = self.links.into_iter().zip(self.blocksizes.into_iter());
+            (
+                FileContent::Spread(RangeLinks::from_links_and_blocksizes(
+                    zipped,
+                    Some(self.offset),
+                )),
+                traversal,
+            )
+        }
+    }
+}
+
+/// Carrier of validation data used between blocks during a walk on the merkle tree.
+#[derive(Debug)]
+pub struct Traversal {
+    last_had_links: bool,
+    last_ending: Ending,
+    last_offset: u64,
+
+    metadata: FileMetadata,
+}
+
+impl Traversal {
+    /// Continues the walk on the merkle tree with the given block contents. The block contents is
+    /// not validated and the range is expected to be the next from previous call to
+    /// FileContent::Spread iterator.
+    ///
+    /// When calling this directly, it is good to note that repeatedly calling this with the same
+    /// block contents will not be detected, and will instead grow the internal Vec of links until
+    /// memory runs out.
+    pub fn continue_walk<'a>(
+        self,
+        next_block: &'a [u8],
+        tree_range: &Range<u64>,
+    ) -> Result<FileReader<'a>, FileReadFailed> {
+        self.last_ending
+            .check_is_suitable_next(self.last_offset, &tree_range)?;
```

Comment:
the `&` shouldn't be needed here

Reply:
Thanks, it was a left over, this didn't always take the range as borrow. Suprisingly no warning on it.